File: frontend/stage13-harness.mjs
import path from "node:path";
import { tmpdir } from "node:os";

const profiles = Object.freeze({
  impl: Object.freeze({
    frontendOrigin: "http://127.0.0.1:3080",
    backendOrigin: "http://127.0.0.1:8080",
    distDir: ".next-stage13-impl",
    tempPrefix: "story-stage13-impl-",
    accountPrefix: "stage13impl",
  }),
  pm3: Object.freeze({
    frontendOrigin: "http://127.0.0.1:3081",
    backendOrigin: "http://127.0.0.1:8081",
    distDir: ".next-stage13-pm3",
    tempPrefix: "story-stage13-pm3-",
    accountPrefix: "stage13pm3",
  }),
  v4impl: Object.freeze({
    frontendOrigin: "http://127.0.0.1:3084",
    backendOrigin: "http://127.0.0.1:8084",
    distDir: ".next-stage13-v4-impl",
    tempPrefix: "story-stage13-v4-impl-",
    accountPrefix: "stage13v4impl",
  }),
  v4pm3: Object.freeze({
    frontendOrigin: "http://127.0.0.1:3085",
    backendOrigin: "http://127.0.0.1:8085",
    distDir: ".next-stage13-v4-pm3",
    tempPrefix: "story-stage13-v4-pm3-",
    accountPrefix: "stage13v4pm3",
  }),
});

function isolatedPath(value, profile, name, { root } = {}) {
  if (!value) throw new Error(`${name}_REQUIRED`);
  const resolved = path.resolve(value);
  const relativeToTemp = path.relative(path.resolve(tmpdir()), resolved);
  if (!relativeToTemp || relativeToTemp.startsWith("..") || path.isAbsolute(relativeToTemp)) {
    throw new Error(`${name}_OUTSIDE_SYSTEM_TEMP`);
  }
  if (!resolved.split(path.sep).some((part) => part.startsWith(profile.tempPrefix))) {
    throw new Error(`${name}_PROFILE_MISMATCH`);
  }
  if (root) {
    const relativeToRoot = path.relative(root, resolved);
    if (relativeToRoot.startsWith("..") || path.isAbsolute(relativeToRoot)) {
      throw new Error(`${name}_OUTSIDE_TEST_ROOT`);
    }
  }
  return resolved;
}

export function validateStage13Harness(env) {
  const profileName = env.STAGE13_HARNESS_PROFILE;
  const profile = profiles[profileName];
  if (!profile) throw new Error("STAGE13_HARNESS_PROFILE_INVALID");
  const exact = {
    E2E_BASE_URL: profile.frontendOrigin,
    E2E_BACKEND_ORIGIN: profile.backendOrigin,
    BACKEND_ORIGIN: profile.backendOrigin,
    PUBLIC_APP_MODE: "0",
    PUBLIC_BASE_URL: profile.frontendOrigin,
    NEXT_DIST_DIR: profile.distDir,
  };
  for (const [name, expected] of Object.entries(exact)) {
    if (env[name] !== expected) throw new Error(`${name}_PROFILE_MISMATCH`);
  }
  if (!env.E2E_ACCOUNT_PREFIX?.startsWith(profile.accountPrefix)) {
    throw new Error("E2E_ACCOUNT_PREFIX_PROFILE_MISMATCH");
  }
  const testRoot = isolatedPath(env.E2E_TEST_ROOT, profile, "E2E_TEST_ROOT");
  const outputDir = isolatedPath(env.E2E_OUTPUT_DIR, profile, "E2E_OUTPUT_DIR", { root: testRoot });
  return {
    profileName,
    ...profile,
    frontendPort: Number(new URL(profile.frontendOrigin).port),
    backendPort: Number(new URL(profile.backendOrigin).port),
    accountPrefix: env.E2E_ACCOUNT_PREFIX,
    testRoot,
    outputDir,
  };
}

File: backend/tests/stage13_harness.py
from __future__ import annotations

import os
from pathlib import Path
import tempfile
from typing import Mapping


PROFILES = {
    "impl": {
        "frontend_origin": "http://127.0.0.1:3080",
        "backend_origin": "http://127.0.0.1:8080",
        "dist_dir": ".next-stage13-impl",
        "temp_prefix": "story-stage13-impl-",
        "account_prefix": "stage13impl",
        "frontend_port": 3080,
        "backend_port": 8080,
    },
    "pm3": {
        "frontend_origin": "http://127.0.0.1:3081",
        "backend_origin": "http://127.0.0.1:8081",
        "dist_dir": ".next-stage13-pm3",
        "temp_prefix": "story-stage13-pm3-",
        "account_prefix": "stage13pm3",
        "frontend_port": 3081,
        "backend_port": 8081,
    },
    "v4impl": {
        "frontend_origin": "http://127.0.0.1:3084",
        "backend_origin": "http://127.0.0.1:8084",
        "dist_dir": ".next-stage13-v4-impl",
        "temp_prefix": "story-stage13-v4-impl-",
        "account_prefix": "stage13v4impl",
        "frontend_port": 3084,
        "backend_port": 8084,
    },
    "v4pm3": {
        "frontend_origin": "http://127.0.0.1:3085",
        "backend_origin": "http://127.0.0.1:8085",
        "dist_dir": ".next-stage13-v4-pm3",
        "temp_prefix": "story-stage13-v4-pm3-",
        "account_prefix": "stage13v4pm3",
        "frontend_port": 3085,
        "backend_port": 8085,
    },
    "v110impl": {
        "frontend_origin": "http://127.0.0.1:3190",
        "backend_origin": "http://127.0.0.1:8190",
        "dist_dir": ".next-v110-impl",
        "temp_prefix": "story-v110-impl-",
        "account_prefix": "v110impl",
        "frontend_port": 3190,
        "backend_port": 8190,
    },
}


def _isolated_path(value: str | None, profile: dict, name: str, root: Path | None = None) -> Path:
    if not value:
        raise RuntimeError(f"{name.lower()}_required")
    resolved = Path(value).resolve()
    system_temp = Path(tempfile.gettempdir()).resolve()
    if system_temp not in resolved.parents or not any(part.startswith(profile["temp_prefix"]) for part in resolved.parts):
        raise RuntimeError(f"{name.lower()}_profile_mismatch")
    if root is not None and resolved != root and root not in resolved.parents:
        raise RuntimeError(f"{name.lower()}_outside_test_root")
    return resolved


def validate_stage13_harness(env: Mapping[str, str] | None = None) -> dict:
    source = os.environ if env is None else env
    profile_name = source.get("STAGE13_HARNESS_PROFILE")
    profile = PROFILES.get(profile_name)
    if profile is None:
        raise RuntimeError("stage13_harness_profile_invalid")
    exact = {
        "E2E_BASE_URL": profile["frontend_origin"],
        "E2E_BACKEND_ORIGIN": profile["backend_origin"],
        "BACKEND_ORIGIN": profile["backend_origin"],
        "PUBLIC_APP_MODE": "0",
        "PUBLIC_BASE_URL": profile["frontend_origin"],
        "NEXT_DIST_DIR": profile["dist_dir"],
    }
    for name, expected in exact.items():
        if source.get(name) != expected:
            raise RuntimeError(f"{name.lower()}_profile_mismatch")
    account_prefix = source.get("E2E_ACCOUNT_PREFIX", "")
    if not account_prefix.startswith(profile["account_prefix"]):
        raise RuntimeError("e2e_account_prefix_profile_mismatch")
    test_root = _isolated_path(source.get("E2E_TEST_ROOT"), profile, "E2E_TEST_ROOT")
    output_dir = _isolated_path(source.get("E2E_OUTPUT_DIR"), profile, "E2E_OUTPUT_DIR", test_root)
    return {"profile_name": profile_name, "test_root": test_root, "output_dir": output_dir, "account_prefix": account_prefix, **profile}
